dbms/src/Storages/DeltaMerge/File/DMFileWithVectorIndexBlockInputStream.cpp

```diff
@@ -227,46 +135,74 @@ void DMFileWithVectorIndexBlockInputStream::internalLoad()
     for (const auto & row : search_results)
         sorted_results.push_back(row.key);
 
-    updateRSResult();
+    updateReadBlockInfos();
 }
 
-void DMFileWithVectorIndexBlockInputStream::updateRSResult()
+void DMFileWithVectorIndexBlockInputStream::updateReadBlockInfos()
 {
     // Vector index is very likely to filter out some packs. For example,
     // if we query for Top 1, then only 1 pack will be remained. So we
-    // update the pack filter used by the DMFileReader to avoid reading
-    // unnecessary data for other columns.
+    // update the reader's read_block_infos to avoid reading unnecessary data for other columns.
+
+    // The following logic is nearly the same with DMFileReader::initReadBlockInfos.
+
+    reader.read_block_infos.clear();
     const auto & pack_stats = dmfile->getPackStats();
-    auto & pack_res = reader.pack_filter.getPackRes();
+    const auto & pack_res = reader.pack_filter.getPackResConst();
 
-    auto results_it = sorted_results.begin();
-    UInt32 pack_start = 0;
-    for (size_t pack_id = 0; pack_id < dmfile->getPacks(); ++pack_id)
-    {
-        if (pack_res[pack_id].isUse())
-            ++valid_packs_before_search;
+    for (const auto res : pack_res)
+        valid_packs_before_search += res.isUse();
 
-        bool pack_has_result = false;
-        UInt32 pack_end = pack_start + pack_stats[pack_id].rows;
-        while (results_it != sorted_results.end() && *results_it >= pack_start && *results_it < pack_end)
+    size_t start_pack_id = 0;
+    size_t read_rows = 0;
+    auto last_pack_res = RSResult::All;
+    auto sorted_results_it = sorted_results.cbegin();
+    size_t pack_id = 0;
+    for (; pack_id < pack_stats.size(); ++pack_id)
+    {
+        if (sorted_results_it == sorted_results.cend())
+            break;
+        auto begin = std::lower_bound(sorted_results_it, sorted_results.cend(), reader.pack_offset[pack_id]);
+        auto end
+            = std::lower_bound(begin, sorted_results.cend(), reader.pack_offset[pack_id] + pack_stats[pack_id].rows);
+        bool is_use = begin != end;
+        bool reach_limit = read_rows >= reader.rows_threshold_per_read;
+        bool break_all_match = !pack_res[pack_id].allMatch() && read_rows >= reader.rows_threshold_per_read / 2;
+
+        if (!is_use)
         {
-            pack_has_result = true;
-            ++results_it;
+            if (read_rows > 0)
+            {
+                reader.read_block_infos.emplace_back(start_pack_id, pack_id - start_pack_id, last_pack_res, read_rows);
+                valid_packs_after_search += (pack_id - start_pack_id);
+            }
+            start_pack_id = pack_id + 1;
+            read_rows = 0;
+            last_pack_res = RSResult::All;
+        }
+        else if (reach_limit || break_all_match)
+        {
+            if (read_rows > 0)
+                reader.read_block_infos.emplace_back(start_pack_id, pack_id - start_pack_id, last_pack_res, read_rows);
```

Comment:
```suggestion
        else if (reach_limit || break_all_match)
        {
            if (read_rows > 0)
                reader.read_block_infos.emplace_back(start_pack_id, pack_id - start_pack_id, last_pack_res, read_rows);
```
Why `valid_packs_after_search` don't need to be updated in this branch?

Reply:
fixed